### ISSUE TITLE
19.13 DDLWorker: Unexpected error, will terminate:: Code: 49

### ISSUE THREAD

**Issue by SaltTan** [ClickHouse/ClickHouse#6639]
I got this with 19.13.3.26.
The story is the same as in #5064
When CH restarts, it starts the DDLWorker cleanup thread, which seems to be triggering a cleanup on all nodes of the cluster.
Another DDLWorker thread is checking and processing the DDL queue, and it runs into 'No node' exceptions. This time DDLWorker seemed to recover nicely after the 'No node' exception, but then it crashed on the next task anyway.

The restarting node:
```
2019.08.23 12:40:17.360397 [ 23 ] {} <Debug> DDLWorker: Processing tasks
2019.08.23 12:40:17.360419 [ 22 ] {} <Debug> DDLWorker: Cleaning queue
```
The other nodes:
```
2019.08.23 12:40:21.166565 [ 24 ] {} <Information> DDLWorker: Task query-0000045304 should be deleted, but it is locked. Skipping it.
2019.08.23 12:40:21.164558 [ 20 ] {} <Information> DDLWorker: Task query-0000045304 should be deleted, but it is locked. Skipping it.
2019.08.23 12:40:21.177815 [ 24 ] {} <Information> DDLWorker: An error occured while checking and cleaning task query-0000045304 from queue: Code: 999, e.displayText() = Coordination::Exception: No node (version 19.13.3.26 (official build))
```
The restarting node:
```
2019.08.23 12:40:21.258063 [ 23 ] {} <Debug> DDLWorker: Processing task query-0000045304 (CREATE TABLE IF NOT EXISTS...)
2019.08.23 12:40:21.558935 [ 23 ] {} <Warning> DDLWorker: An error occurred while processing task query-0000045304 (CREATE TABLE IF NOT EXISTS...) : Code: 999, e.displayText() = Coordination::Exception: No node, Stack trace:

0. clickhouse-server(StackTrace::StackTrace()+0x30) [0x7f6b5f0]
1. clickhouse-server(Coordination::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)+0x30) [0x764f210]
2. clickhouse-server(Coordination::Exception::Exception(int)+0xa7) [0x764f367]
3. clickhouse-server() [0x3964de2]
4. clickhouse-server(DB::DDLWorker::processTask(DB::DDLTask&, std::shared_ptr<zkutil::ZooKeeper> const&)+0x203) [0x6d20ba3]
5. clickhouse-server(DB::DDLWorker::processTasks()+0x5d6) [0x6d21b36]
6. clickhouse-server(DB::DDLWorker::runMainThread()+0x294) [0x6d227f4]
7. clickhouse-server(ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::DDLWorker::*)(), DB::DDLWorker*>(void (DB::DDLWorker::*&&)(), DB::DDLWorker*&&)::{lambda()#1}::operator()() const+0x4b) [0x6d2486b]
8. clickhouse-server(ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>)+0x1a6) [0x3c760f6]
9. clickhouse-server() [0xba3e1a0]
10. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7f9e1a213494]
11. /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f9e19a4dacf]
 (version 19.13.3.26 (official build))
2019.08.23 12:40:21.559010 [ 23 ] {} <Error> DDLWorker: ZooKeeper error: Code: 999, e.displayText() = Coordination::Exception: No node, Stack trace:

0. clickhouse-server(StackTrace::StackTrace()+0x30) [0x7f6b5f0]
1. clickhouse-server(Coordination::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int)+0x30) [0x764f210]
2. clickhouse-server(Coordination::Exception::Exception(int)+0xa7) [0x764f367]
3. clickhouse-server() [0x3964de2]
4. clickhouse-server(DB::DDLWorker::processTask(DB::DDLTask&, std::shared_ptr<zkutil::ZooKeeper> const&)+0x203) [0x6d20ba3]
5. clickhouse-server(DB::DDLWorker::processTasks()+0x5d6) [0x6d21b36]
6. clickhouse-server(DB::DDLWorker::runMainThread()+0x294) [0x6d227f4]
7. clickhouse-server(ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::DDLWorker::*)(), DB::DDLWorker*>(void (DB::DDLWorker::*&&)(), DB::DDLWorker*&&)::{lambda()#1}::operator()() const+0x4b) [0x6d2486b]
8. clickhouse-server(ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>)+0x1a6) [0x3c760f6]
9. clickhouse-server() [0xba3e1a0]
10. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7f9e1a213494]
11. /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f9e19a4dacf]
 (version 19.13.3.26 (official build))

2019.08.23 12:40:21.559019 [ 23 ] {} <Debug> DDLWorker: Processing tasks
2019.08.23 12:40:21.683882 [ 23 ] {} <Information> DDLWorker: Task query-0000045304 was deleted from ZooKeeper before current host committed it

2019.08.23 12:40:21.949695 [ 23 ] {} <Error> DDLWorker: Unexpected error, will terminate:: Code: 49, e.displayText() = DB::Exception: Server expects that DDL task query-0000045317 should be processed, but it was already processed according to ZK, Stack trace:

0. clickhouse-server(StackTrace::StackTrace()+0x30) [0x7f6b5f0]
1. clickhouse-server(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x25) [0x3c46865]
2. clickhouse-server() [0x396584a]
3. clickhouse-server(DB::DDLWorker::runMainThread()+0x294) [0x6d227f4]
4. clickhouse-server(ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::DDLWorker::*)(), DB::DDLWorker*>(void (DB::DDLWorker::*&&)(), DB::DDLWorker*&&)::{lambda()#1}::operator()() const+0x4b) [0x6d2486b]
5. clickhouse-server(ThreadPoolImpl<std::thread>::worker(std::_List_iterator<std::thread>)+0x1a6) [0x3c760f6]
6. clickhouse-server() [0xba3e1a0]
7. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7f9e1a213494]
8. /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f9e19a4dacf]
 (version 19.13.3.26 (official build))
```
After another restart query-0000045317 was processed without issues.

**Comment by Phantomape** [comment]
Able to reproduce in our 4-node servers via Clickhouse 19.16.2.2. It seems like a race condition to me, and we bypass these errors by letting the two Poco Event variable in DDLWorker to send continue signal even though these two exceptions happen. However, that may not seem like a correct fix at all, more like a dirty hack.

**Comment by alexey-milovidov** [comment]
There is no such error message in recent ClickHouse versions:

```
milovidov@milovidov-desktop:~/work/ClickHouse/src$ rg 'will terminate'

```

It means that the relevant code has been reworked and there is a high change the issue is fixed.

**Comment by alexey-milovidov** [comment]
Fixed here: #18285
